src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1818,7 +1818,7 @@ InsertPlaceholderCoordinatorRecord(void)
 	bool nodeAlreadyExists = false;
 
 	/* as long as there is a single node, localhost should be ok */
-	AddNodeMetadata("localhost", PostPortNumber, &nodeMetadata, &nodeAlreadyExists);
+	AddNodeMetadata(LocalHostName, PostPortNumber, &nodeMetadata, &nodeAlreadyExists);
```

Comment:
Still, here is a little tricky . If the user changes `citus.local_hostname ` later on, the `pg_dist_node` entry would be stale. 

Also, in a sense, both `citus.local_hostname` and `citus_set_coordinator_host `/`citus_update_node` are capable of changing the hostnames to the connections to the local host -- on the coordinator. They can diverge and even the same query execution may use both addresses :)

Do you have any suggestions to avoid this? (I have some ideas, but they don't sound simple enough so let me not steer your ideas at this point)

Reply:
> Shouldn't we still need to change the two hardcoded localhost uses to LocalHostName in AddNodeMetadata()? Because assume that the first thing you do in the cluster is to set LocalHostName. Later, create a distributed table. Then, coordinatorNode->workerName would be LocalHostName and the logic in AddNodeMetadata() is broken.

The two hardcoded `"localhost"` occurrences there seem to be correct to me (we could use the define though).

Both lead towards the error that you can't add worker nodes as long as the coordinator refers to itself as `"localhost"`. If we would compare that to the configured hostname it would cause the localhost error to be returned even though the hostname of the coordinator uses a good hostname.

I am not super happy with the error checks we perform, they are very narrow and only cover the `"localhost"` case. IMHO we should aim to remove that error and have actual checks from the worker we are adding if they can reach the coordinator (and other workers) based on the configuration. We could then provide good errors to the user as `the worker you are adding is unable to connect to the coordinator with its current configured hostname: localhost`, etc.

> I made the hostname GUC a super user setting which requires a reload. It doesn't seem added as superuser GUC?

I will need to verify this, might be an oversight I not made this a super user setting.

Reply:
> the worker you are adding is unable to connect to the coordinator with its current configured hostname: localhost, etc.

Agree, I think Marco might have opened an issue on this as well.



Alright, I think there is a minor difference on our perspectives. My understanding from your comments is that once the GUC is set, it is going to be set to a hostname that can be accessed via the network. So, everyone could actually connect to that hostname afterwards.

 But, it is also possible that you just use hostname like `MacBook-Pro-Onders.local` for the GUC.

I think it is probably unlikely to set something like `MacBook-Pro-Onders.local` and that's why you think it is not necessary to change the hard coded values.

I think you are probably right, though I couldn't follow this:

> If we would compare that to the configured hostname it would cause the localhost error to be returned even though the hostname of the coordinator uses a good hostname.

My suggestion is to compared `hostName == 'localhost' OR hostname == LocalHostName`. How would that throw errors? Still, my suggestion also wouldn't work if the user changes `LocalHostName` multiple times.

So, I think I'm fine with the current approach for now

Reply:
For the `AddNodeMetadata` function with the hardcoded `"localhost"` we talk about this section: https://github.com/citusdata/citus/blob/acdde9fa2de2cda749d8b1654c52734148a93401/src/backend/distributed/metadata/node_metadata.c#L1420-L1441

The purpose of the checks is to decide to _not_ allow the user to add a node when the coordinator is added via the `"localhost"` hostname as we 'assume' (probably correctly) that that hostname will not resolve to the coordinator from the worker we are adding. By disallowing the user to add a node when this happen we hope to trigger the user to set a network routable hostname for the coordinator.

The goal of this GUC is to use the network routable hostname during connections locally. This to allow users to configure ssl certificates with that hostname in the CN or ALT names of the certificate so that `verify-full` can be used on the ssl connections.

Disallowing users to add a new node when the coordinator is configured with a hostname equal to the value of `LocalHostName` would prevent this from working.

For hostnames other than `"localhost"` that would only correctly resolve from the current node this will not change much, see the suggestion earlier where we actually try to connect from the worker back to the coordinator with the configured hostname.